src/porepy/numerics/ad/secondary_operator.py

```diff
@@ -0,0 +1,897 @@
+        which is time step and iterate index dependent.
+
+        This function is assigned by the factory class :class:`SecondaryExpression`
+        which has a reference to the md-grid and must not be touched by the user.
+
+        """
+
+    def __repr__(self) -> str:
+        """String representation giving information on name, time and iterate index, as
+        well as domains and dependencies."""
+
+        msg = (
+            f"Secondary operator {self.name}.\n"
+            + f"\nDefined on {len(self._domains)} {self._domain_type}.\n"
+            + f"Dependent on {len(self.children)} independent expressions.\n"
+        )
+
+        if self.is_previous_time:
+            msg += f"Evaluated at the previous time step {self.time_step_index}.\n"
+        elif self.is_previous_iterate:
+            msg += f"Evaluated at the previous iterate {self.iterate_index}.\n"
+
+        return msg
+
+    def __call__(self, *args: Operator) -> Operator:
+        """By calling this operator with new children, the user can change the
+        structure of the Jacobian.
+
+        The dependencies/children are replaced and hence the non-trivial entries in
+        the Jacobian in the md-setting.
+
+        Note:
+            Call this operator only if you know what you are doing.
+
+            It will reset the time step and iterate indices to current time and iterate.
+
+        """
+        op = SecondaryOperator(
+            name=self.name,
+            domains=self.domains,  # type:ignore[arg-type]
+            children=args,  # type:ignore[arg-type]
+        )
+        op.fetch_data = self.fetch_data
+        return op
+
+    def previous_timestep(self, steps: int = 1) -> SecondaryOperator:
+        """Secondary operators have children which also need to be obtained at
+        the previous time step."""
+
+        op = super().previous_timestep(steps=steps)
+        op.children = [child.previous_timestep(steps=steps) for child in self.children]
+        return op
+
+    def previous_iteration(self, steps: int = 1) -> SecondaryOperator:
+        """Secondary operators have children which also need to be obtained at
+        the previous iteration."""
+        op = super().previous_iteration(steps=steps)
+        op.children = [child.previous_iteration(steps=steps) for child in self.children]
+        return op
+
+    def func(self, *args: FloatType) -> float | np.ndarray | AdArray:
+        """See :meth:`~porepy.numerics.ad.operator_functions.AbstractFunction.func`.
+
+        Note:
+            This class cannot inherit from
+            :class:`~porepy.numerics.ad.operator_functions.AbstractFunction` because
+            that parent would block arithmetic overloads.
+            But it uses the operation ``evaluate`` and the ``AbstractFunction.func``
+            with ``self`` as an explicit argument.
+
+        """
+        # mypy complains that self is not an instance of AbstractFunction because of the
+        # hack
```

Comment:
I would require here the same code as implemented in `AbstractFunction` to call `get_values` and `get_jacobian`, this "hack" avoids code duplicates.
It is a "hack" because it does not inherit from `AbstractFunction` because of the blocking of arithmetic overloads.

I could resolve this by refactoring the `AbstractFuncion` to be a factory as intended, and not an operator.

Reply:
Let's not complicate this PR by changing other parts of the code.

Please be a bit more specific on what is meant by 'the hack', to avoid someone being confused at a later stage.

Reply:
Added better documentation.